File: src/planner/calculations.rs
use std::collections::HashMap;

use crate::models::Food;
use crate::planner::constants::*;

/// Nutrient density breakdown.
#[derive(Debug, Clone, Default)]
pub struct NutrientDensity {
    pub carbs: f64,
    pub protein: f64,
    pub fats: f64,
    pub vitamins: f64,
}

impl NutrientDensity {
    /// Sum of all nutrient densities.
    pub fn sum(&self) -> f64 {
        self.carbs + self.protein + self.fats + self.vitamins
    }

    /// Minimum non-zero nutrient value.
    pub fn min_nonzero(&self) -> f64 {
        [self.carbs, self.protein, self.fats, self.vitamins]
            .into_iter()
            .filter(|&v| v > 0.0)
            .fold(f64::MAX, f64::min)
    }

    /// Maximum nutrient value.
    pub fn max(&self) -> f64 {
        [self.carbs, self.protein, self.fats, self.vitamins]
            .into_iter()
            .fold(0.0, f64::max)
    }
}

/// Calculate calorie-weighted nutrient density from stomach contents.
///
/// Returns (density, total_calories).
pub fn sum_all_weighted_nutrients(stomach: &HashMap<&Food, u32>) -> (NutrientDensity, f64) {
    let total_cal: f64 = stomach
        .iter()
        .map(|(f, qty)| f.calories * (*qty) as f64)
        .sum();

    if total_cal == 0.0 {
        return (NutrientDensity::default(), 0.0);
    }

    let mut density = NutrientDensity::default();
    for (food, qty) in stomach {
        let weight = (food.calories * (*qty) as f64) / total_cal;
        density.carbs += food.carbs * weight;
        density.protein += food.protein * weight;
        density.fats += food.fats * weight;
        density.vitamins += food.vitamins * weight;
    }

    (density, total_cal)
}

/// Calculate balance multiplier.
///
/// Range: 0.5 (worst balance) to 2.0 (perfect 1:1:1:1 ratio).
pub fn calculate_balance_mult(density: &NutrientDensity) -> f64 {
    let max_val = density.max();
    if max_val == 0.0 {
        return 1.0; // No food = neutral
    }

    let min_val = density.min_nonzero();
    if min_val == f64::MAX {
        return BALANCE_MULT_MIN; // Missing nutrients = worst
    }

    let balance_ratio = min_val / max_val; // 0.0 to 1.0
                                           // Map ratio to multiplier range: 0 -> 0.5, 1 -> 2.0
    BALANCE_MULT_MIN + balance_ratio * (BALANCE_MULT_MAX - BALANCE_MULT_MIN)
}

/// Calculate variety multiplier.
///
/// Uses exponential approach: each +20 qualifying foods halves remaining gap.
/// Range: 1.0 (no variety) to VARIETY_MULT_MAX (1.55).
pub fn calculate_variety_mult(variety_count: usize) -> f64 {
    // variety_mult(count) = 1.0 + (VARIETY_MULT_MAX - 1.0) * (1 - 0.5^(count/20))
    let bonus_range = VARIETY_MULT_MAX - 1.0;
    1.0 + bonus_range * (1.0 - 0.5_f64.powf(variety_count as f64 / 20.0))
}

/// Calculate taste multiplier.
///
/// Calorie-weighted average of individual food taste multipliers.
/// Range: 0.7 (all hated) to 1.3 (all favorite).
pub fn calculate_taste_mult(stomach: &HashMap<&Food, u32>) -> f64 {
    let total_cal: f64 = stomach
        .iter()
        .map(|(f, qty)| f.calories * (*qty) as f64)
        .sum();

    if total_cal == 0.0 {
        return 1.0; // No food = neutral
    }

    let weighted_taste: f64 = stomach
        .iter()
        .map(|(food, qty)| {
            let cal = food.calories * (*qty) as f64;
            let mult = tastiness_multiplier(food.tastiness);
            cal * mult
        })
        .sum();

    weighted_taste / total_cal
}

/// Count foods that qualify for variety bonus.
pub fn count_variety_qualifying(stomach: &HashMap<&Food, u32>) -> usize {
    stomach
        .iter()
        .filter(|(food, qty)| is_variety_qualifying(food.calories, **qty))
        .count()
}

/// Check if a food qualifies for variety bonus.
pub fn is_variety_qualifying(calories_per_unit: f64, quantity: u32) -> bool {
    (calories_per_unit * quantity as f64) >= VARIETY_CAL_THRESHOLD
}

/// Calculate craving multiplier.
///
/// Returns 1.0 + (CRAVING_MULT_PER_MATCH * matches) for foods matching cravings.
pub fn calculate_craving_mult(stomach: &HashMap<&Food, u32>, cravings: &[String]) -> f64 {
    let craving_set: std::collections::HashSet<String> =
        cravings.iter().map(|c| c.to_lowercase()).collect();

    let matches = stomach
        .keys()
        .filter(|f| craving_set.contains(&f.name.to_lowercase()))
        .count();

    1.0 + matches as f64 * CRAVING_MULT_PER_MATCH
}

/// Configurable multipliers for SP calculation.
#[derive(Debug, Clone)]
pub struct SpConfig {
    pub server_mult: f64,
    pub dinner_party_mult: f64,
}

impl Default for SpConfig {
    fn default() -> Self {
        Self {
            server_mult: DEFAULT_SERVER_MULT,
            dinner_party_mult: DEFAULT_DINNER_PARTY_MULT,
        }
    }
}

/// Calculate total SP from stomach contents and craving state.
///
/// Formula: (nutrient_total * balance * variety * taste * craving * dinner_party + base) * server
pub fn calculate_sp(stomach: &HashMap<&Food, u32>, cravings: &[String], config: &SpConfig) -> f64 {
    let (density, _total_cal) = sum_all_weighted_nutrients(stomach);
    let density_sum = density.sum();

    let balance_mult = calculate_balance_mult(&density);
    let variety_count = count_variety_qualifying(stomach);
    let variety_mult = calculate_variety_mult(variety_count);
    let taste_mult = calculate_taste_mult(stomach);
    let craving_mult = calculate_craving_mult(stomach, cravings);

    // Chain multipliers
    let nutrition_sp = density_sum
        * balance_mult
        * variety_mult
        * taste_mult
        * craving_mult
        * config.dinner_party_mult;

    (nutrition_sp + BASE_SKILL_POINTS) * config.server_mult
}

/// Calculate the SP delta from adding one unit of a food.
pub fn get_sp_delta(
    stomach: &HashMap<&Food, u32>,
    food: &Food,
    cravings: &[String],
    config: &SpConfig,
) -> f64 {
    let sp_before = calculate_sp(stomach, cravings, config);

    // Create new stomach with food added
    let mut new_stomach = stomach.clone();
    let current = new_stomach.get(&food).copied().unwrap_or(0);
    new_stomach.insert(food, current + 1);

    let sp_after = calculate_sp(&new_stomach, cravings, config);

    sp_after - sp_before
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_food(name: &str, cal: f64, c: f64, p: f64, f: f64, v: f64, taste: i8) -> Food {
        Food {
            name: name.to_string(),
            calories: cal,
            carbs: c,
            protein: p,
            fats: f,
            vitamins: v,
            tastiness: taste,
            stomach: 0,
            available: 10,
        }
    }

    #[test]
    fn test_balance_mult_perfect() {
        // Perfect balance: all nutrients equal -> 2.0x
        let density = NutrientDensity {
            carbs: 10.0,
            protein: 10.0,
            fats: 10.0,
            vitamins: 10.0,
        };
        let mult = calculate_balance_mult(&density);
        assert!((mult - BALANCE_MULT_MAX).abs() < 0.01);
    }

    #[test]
    fn test_balance_mult_imbalanced() {
        // Very imbalanced: ratio = 0.1 -> 0.5 + 0.1 * 1.5 = 0.65
        let density = NutrientDensity {
            carbs: 100.0,
            protein: 10.0,
            fats: 10.0,
            vitamins: 10.0,
        };
        let mult = calculate_balance_mult(&density);
        let expected = BALANCE_MULT_MIN + 0.1 * (BALANCE_MULT_MAX - BALANCE_MULT_MIN);
        assert!((mult - expected).abs() < 0.01);
    }

    #[test]
    fn test_variety_mult() {
        assert!((calculate_variety_mult(0) - 1.0).abs() < 0.01); // No variety = 1.0x
                                                                 // 20 foods = halfway to max
        let half_bonus = 1.0 + (VARIETY_MULT_MAX - 1.0) * 0.5;
        assert!((calculate_variety_mult(20) - half_bonus).abs() < 0.01);
        assert!(calculate_variety_mult(100) < VARIETY_MULT_MAX); // Approaches but never exceeds
    }

    #[test]
    fn test_taste_mult() {
        let food1 = sample_food("Good", 100.0, 10.0, 10.0, 10.0, 10.0, 3); // favorite
        let food2 = sample_food("Bad", 100.0, 10.0, 10.0, 10.0, 10.0, -3); // hated

        let mut stomach: HashMap<&Food, u32> = HashMap::new();
        stomach.insert(&food1, 1);
        let mult = calculate_taste_mult(&stomach);
        assert!((mult - 1.3).abs() < 0.01); // +3 taste = 1.3x

        let mut stomach2: HashMap<&Food, u32> = HashMap::new();
        stomach2.insert(&food2, 1);
        let mult2 = calculate_taste_mult(&stomach2);
        assert!((mult2 - 0.7).abs() < 0.01); // -3 taste = 0.7x
    }

    #[test]
    fn test_is_variety_qualifying() {
        assert!(is_variety_qualifying(500.0, 4)); // 2000 cal
        assert!(!is_variety_qualifying(500.0, 3)); // 1500 cal
        assert!(is_variety_qualifying(2000.0, 1)); // Exactly 2000
    }

    #[test]
    fn test_calculate_sp_empty_stomach() {
        let stomach: HashMap<&Food, u32> = HashMap::new();
        let config = SpConfig::default();
        let sp = calculate_sp(&stomach, &[], &config);
        assert!((sp - BASE_SKILL_POINTS).abs() < 0.01);
    }

    #[test]
    fn test_sp_delta() {
        let food = sample_food("Balanced", 500.0, 10.0, 10.0, 10.0, 10.0, 2);
        let stomach: HashMap<&Food, u32> = HashMap::new();
        let config = SpConfig::default();

        let delta = get_sp_delta(&stomach, &food, &[], &config);
        assert!(delta > 0.0); // Adding food should increase SP
    }

    #[test]
    fn test_craving_mult() {
        let food = sample_food("Pizza", 500.0, 10.0, 10.0, 10.0, 10.0, 2);
        let mut stomach: HashMap<&Food, u32> = HashMap::new();
        stomach.insert(&food, 1);

        // No craving = 1.0x
        let mult_none = calculate_craving_mult(&stomach, &[]);
        assert!((mult_none - 1.0).abs() < 0.01);

        // Matching craving = 1.0 + CRAVING_MULT_PER_MATCH
        let mult_match = calculate_craving_mult(&stomach, &["Pizza".to_string()]);
        assert!((mult_match - (1.0 + CRAVING_MULT_PER_MATCH)).abs() < 0.01);
    }

    #[test]
    fn test_server_mult_affects_sp() {
        let food = sample_food("Test", 500.0, 10.0, 10.0, 10.0, 10.0, 0);
        let mut stomach: HashMap<&Food, u32> = HashMap::new();
        stomach.insert(&food, 1);

        let config_1x = SpConfig::default();
        let config_2x = SpConfig {
            server_mult: 2.0,
            ..Default::default()
        };

        let sp_1x = calculate_sp(&stomach, &[], &config_1x);
        let sp_2x = calculate_sp(&stomach, &[], &config_2x);

        assert!((sp_2x / sp_1x - 2.0).abs() < 0.01);
    }
}

File: calculations.py
"""Pure scoring utilities for nutrients, bonuses, and SP.

Contains deterministic helpers for nutrient totals, balanced diet/variety/tastiness
bonuses, and SP calculations, plus small simulators used by the planner.

Exports
-------
sum_weighted_nutrients
sum_all_weighted_nutrients
calculate_balanced_diet_bonus
calculate_balanced_diet_ratio
get_tastiness_bonus
calculate_nutrition_multiplier
get_sp
simulate_stomach_with_added_food
evaluate_bonus_with_addition
get_sp_delta
get_balanced_diet_ratio
get_variety_bonus
tastiness_delta_for_added_unit
variety_fraction_for
is_variety_qualifying
soft_variety_count
variety_count

Notes
-----
All functions are side-effect free; inputs are treated as read-only.
"""

import logging
from typing import Dict

from constants import (
    BASE_SKILL_POINTS,
    CRAVING_SATISFIED_FRAC,
    TASTINESS_MULTIPLIERS,
    TASTINESS_WEIGHT,
    VARIETY_BONUS_CAP_PP,
    VARIETY_CAL_THRESHOLD,
)
from models.food import (
    Food,
)

logger = logging.getLogger(__name__)


def _unique_variety_names(
    stomach: dict,
) -> set[str]:
    # only items that individually meet the variety threshold
    return {
        food.name.lower()
        for food, quantity in stomach.items()
        if is_variety_qualifying(food, quantity)
    }


def _total_calories(
    stomach: dict,
) -> float:
    return sum(food.calories * quantity for food, quantity in stomach.items())


def sum_weighted_nutrients(
    stomach,
    attr,
):
    """Calorie-weighted **sum** for a single nutrient.

    Notes
    -----
    Uses the per-food calorie share computed in `sum_all_weighted_nutrients`
    (not raw calories multiplication).

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.
    attr : str
        Nutrient attribute name.
        Examples: "carbs", "protein", "fat", "vitamins".

    Returns
    -------
    float
        Weighted sum for the selected nutrient.
        (Sum over ``nutrient * quantity``.)
    """
    # Weight by calories*quantity so high-cal foods influence balance
    return sum(getattr(food, attr) * quantity for food, quantity in stomach.items())


def sum_all_weighted_nutrients(
    stomach: dict,
):
    """Weighted sums for all nutrient types.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.

    Returns
    -------
    tuple[dict[str, float], float]
        ``(density_dict, total_calories)`` where the dict has keys
        ``"carbs"``, ``"protein"``, ``"fat"``, ``"vitamins"``.
    """
    # Delegate per-attr computation to keep logic in one place
    total_cal = _total_calories(stomach)
    totals = {"carbs": 0.0, "protein": 0.0, "fat": 0.0, "vitamins": 0.0}
    if total_cal == 0:
        return totals, 0.0

    for food, quantity in stomach.items():
        calorie_weight = (food.calories * quantity) / total_cal
        totals["carbs"] += food.carbs * calorie_weight
        totals["protein"] += food.protein * calorie_weight
        totals["fat"] += food.fat * calorie_weight
        totals["vitamins"] += food.vitamins * calorie_weight

    return totals, total_cal


def calculate_balanced_diet_bonus(
    nutrients: list[float],
) -> float:
    """Convert balance ratio to bonus percentage.

    Formula: ``(min_nonzero / max) * 100 - 50``.

    Parameters
    ----------
    nutrients : list of float
        Weighted nutrient totals.

    Returns
    -------
    float
        Bonus percentage in ``[-50, +50]``.
    """

    ratio = calculate_balanced_diet_ratio(nutrients)
    return (ratio * 100) - 50


def calculate_balanced_diet_ratio(
    nutrients: list[float],
) -> float:
    """Balance ratio of the nutrient distribution.

    Defined as ``min_nonzero / max`` among the nutrient values.

    Parameters
    ----------
    nutrients : list of float
        Weighted nutrient totals.

    Returns
    -------
    float
        Ratio in ``[0, 1]``; higher means more balanced.
    """

    # If all totals are zero, ratio is 0; otherwise min/max including zeros
    max_nutrient = max(nutrients)
    if max_nutrient <= 0:
        return 0.0
    return min(nutrients) / max_nutrient


def get_tastiness_bonus(
    stomach: dict,
) -> float:
    """Tastiness bonus percentage points for the current stomach.

    A calorie-weighted average of tastiness multipliers.
    Scaled by ``TASTINESS_WEIGHT``.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.

    Returns
    -------
    float
        Taste bonus in percentage points.
    """

    total_cal = sum(food.calories * quantity for food, quantity in stomach.items())
    if total_cal <= 0:
        return 0.0
    # Map tastiness → multiplier (fraction); default 0 for unknowns.
    # Convert to percentage points below.
    taste_score = sum(
        TASTINESS_MULTIPLIERS.get(food.tastiness, 0.0) * food.calories * quantity
        for food, quantity in stomach.items()
    )
    return (taste_score / total_cal) * 100.0 * TASTINESS_WEIGHT


def calculate_nutrition_multiplier(
    stomach: dict,
    cravings: list[str],
    unique_foods_24h: set[str],
) -> float:
    """Total nutrition multiplier (percentage points).

    Includes balance, variety, and taste bonuses.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.
    cravings : list of str
        Active craving names (case-insensitive).  Reserved for
        future use; not currently consumed by this function.
    unique_foods_24h : set of str
        Foods (names, lowercased) that qualify for variety
        within the 24h window.

    Returns
    -------
    float
        Sum of bonus components in percentage points.
    """

    # ---- balanced diet (ordered list, not dict) ----
    density, _ = sum_all_weighted_nutrients(stomach)

    balanced_diet_pp = calculate_balanced_diet_bonus(
        [
            density["carbs"],
            density["protein"],
            density["fat"],
            density["vitamins"],
        ]
    )
    variety_pp = get_variety_bonus(len(unique_foods_24h))
    tastiness_pp = get_tastiness_bonus(stomach)

    # returns percentage points (not a fraction)
    return balanced_diet_pp + variety_pp + tastiness_pp


def get_sp(
    stomach,
    cravings,
    cravings_satisfied,
    unique_foods_24h,
    *,
    server_mult: float = 1.0,
    dinner_party_mult: float = 1.0,
) -> float:
    """Compute SP (skill points) from stomach and bonuses.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.
    cravings : list of str
        Active craving names.
    cravings_satisfied : int
        Number of cravings already satisfied.
    unique_foods_24h : set of str
        Variety reference set (qualified foods by name).
    server_mult : float, optional
        Server skill gain multiplier. Default is 1.0.
    dinner_party_mult : float, optional
        Dinner party multiplier (1.0-3.0). Default is 1.0.

    Returns
    -------
    float
        Final SP value.

    Notes
    -----
    Formula: ((nutrient_sp * bonuses * dinner_party) + BASE_SP) * server_mult
    """

    density, _ = sum_all_weighted_nutrients(stomach)  # calorie-weighted avg
    density_sum = (
        density["carbs"] + density["protein"] + density["fat"] + density["vitamins"]
    )

    bonus = (
        calculate_nutrition_multiplier(
            stomach,
            cravings,
            unique_foods_24h,
        )
        / 100.0
    )
    bonus += cravings_satisfied * CRAVING_SATISFIED_FRAC

    # Apply dinner party multiplier to nutrition SP, then add base, then server mult
    nutrition_sp = density_sum * (1.0 + bonus) * dinner_party_mult
    return (nutrition_sp + BASE_SKILL_POINTS) * server_mult


def simulate_stomach_with_added_food(
    stomach: dict,
    food_to_add: Food,
) -> dict:
    """Clone the stomach and add one unit of the given food.

    Parameters
    ----------
    stomach : dict[Food, int]
        Original stomach state.
    food_to_add : Food
        Food to add.

    Returns
    -------
    dict[Food, int]
        New stomach mapping with the increment applied.
    """

    # Work on a shallow copy; never mutate the caller's stomach
    clone = dict(stomach)
    clone[food_to_add] = (
        clone.get(
            food_to_add,
            0,
        )
        + 1
    )
    return clone


def evaluate_bonus_with_addition(
    stomach,
    food_to_add,
    cravings,
    variety_reference: set[str],
):
    """Estimate nutrition multiplier after hypothetically adding a food.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.
    food_to_add : Food
        Food to consider adding.
    cravings : list of str
        Active craving names.
    variety_reference : set of str
        Names (lowercased) that already meet the variety threshold.

    Returns
    -------
    float
        Predicted total bonus (percentage points) after the addition.
    """

    test_stomach = simulate_stomach_with_added_food(
        stomach,
        food_to_add,
    )

    # Only add to variety set if this bite newly meets the per-food threshold
    new_total = food_to_add.calories * test_stomach[food_to_add]
    if (
        new_total >= VARIETY_CAL_THRESHOLD
        and food_to_add.name.lower() not in variety_reference
    ):
        updated_variety = variety_reference | {food_to_add.name.lower()}
    else:
        updated_variety = variety_reference

    return calculate_nutrition_multiplier(
        test_stomach,
        cravings,
        updated_variety,
    )


def get_sp_delta(
    food,
    stomach,
    cravings,
    cravings_satisfied,
) -> float:
    """Change in SP from adding one unit of a specific food.

    Parameters
    ----------
    food : Food
        Candidate food to add.
    stomach : dict[Food, int]
        Current stomach state.
    cravings : list of str
        Active craving names.
    cravings_satisfied : int
        Number of cravings already satisfied.

    Returns
    -------
    float
        ``SP(after) - SP(before)`` for a single added unit.
    """

    # Marginal SP: SP(after adding one unit) − SP(before)
    after = simulate_stomach_with_added_food(
        stomach,
        food,
    )
    delta = get_sp(
        after,
        cravings,
        cravings_satisfied,
        _unique_variety_names(after),
    ) - get_sp(
        stomach,
        cravings,
        cravings_satisfied,
        _unique_variety_names(stomach),
    )
    return delta


def get_balanced_diet_ratio(
    stomach,
):
    """Balance ratio for the current stomach.

    Returns
    -------
    float
        ``min_nonzero / max`` using weighted nutrient totals.
    """
    density, _ = sum_all_weighted_nutrients(stomach)
    nutrients = [
        density["carbs"],
        density["protein"],
        density["fat"],
        density["vitamins"],
    ]
    max_nutrient = max(nutrients)
    if max_nutrient <= 0:
        return 0
    return min(nutrients) / max_nutrient


def get_variety_bonus(
    unique_food_count,
):
    """Variety bonus from the number of qualifying foods.

    Parameters
    ----------
    unique_food_count : int
        Count of foods meeting the variety calorie threshold.

    Returns
    -------
    float
        Variety bonus in percentage points (capped).
    """

    # Exponential cap: each +20 qualifying foods halves the remaining gap
    # to the cap
    return VARIETY_BONUS_CAP_PP * (1 - 0.5 ** (unique_food_count / 20))


def tastiness_delta_for_added_unit(
    stomach: dict[Food, int],
    food: Food,
) -> float:
    """Change in tastiness bonus from adding one unit of a food.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.
    food : Food
        Food to add.

    Returns
    -------
    float
        ``tastiness_bonus(after) - tastiness_bonus(before)`` in percentage points.
    """

    before = get_tastiness_bonus(stomach)
    after = get_tastiness_bonus(
        simulate_stomach_with_added_food(
            stomach,
            food,
        )
    )
    return after - before


def variety_fraction_for(
    food_item: Food,
    quantity: int,
) -> float:
    """Fractional variety contribution of a single food.

    Parameters
    ----------
    food_item : Food
        Food to evaluate.
    quantity : int
        Quantity in the stomach.

    Returns
    -------
    float
        Value in ``[0.0, 1.0]``.
        Computed as ``min(1, calories*quantity / VARIETY_CAL_THRESHOLD)``.
    """

    # No contribution when quantity ≤ 0
    if quantity <= 0:
        return 0.0
    # Fraction toward threshold (clamped to 1.0)
    return min(
        1.0,
        (food_item.calories * quantity) / VARIETY_CAL_THRESHOLD,
    )


def is_variety_qualifying(
    food_item: Food,
    quantity: int,
) -> bool:
    """Whether this food alone meets the variety threshold.

    Parameters
    ----------
    food_item : Food
        Food to evaluate.
    quantity : int
        Quantity in the stomach.

    Returns
    -------
    bool
        ``True`` if ``calories * quantity >= VARIETY_CAL_THRESHOLD``.
    """

    # Qualifies if calories * quantity meets or exceeds the threshold
    return (food_item.calories * quantity) >= VARIETY_CAL_THRESHOLD


def soft_variety_count(
    stomach: Dict[Food, int],
) -> float:
    """Soft (fractional) variety count across all foods.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.

    Returns
    -------
    float
        Sum of fractional contributions (can be non-integer).
    """

    # Sum fractional contributions for soft variety
    return sum(
        variety_fraction_for(food_item, quantity)
        for food_item, quantity in stomach.items()
    )


def variety_count(
    stomach: Dict[Food, int],
) -> int:
    """Count of foods that individually meet the variety threshold.

    Parameters
    ----------
    stomach : dict[Food, int]
        Current stomach state.

    Returns
    -------
    int
        Number of qualifying foods.
    """

    # Count foods that individually meet the threshold (hard variety)
    return sum(
        1
        for food_item, quantity in stomach.items()
        if is_variety_qualifying(food_item, quantity)
    )
